File: 12.py
from functools import reduce

def is_prime_basic(x):
    if x <= 1:
        return False
    elif x <= 3:
        return True
    elif x % 2 == 0 or x % 3 == 0:
        return False
    i = 5
    while i ** 2 <= x:
        if x % i == 0 or x % (i+2) == 0:
            return False
        i += 6
    return True


def prime_factorisation(n, primes):
    factorisation = []
    i = 0
    temp = [0,0]
    while n != 1:
        p = primes[i]
        while n % p == 0:
            temp = [p, temp[1] + 1]
            n = n // p
        factorisation.append(tuple(temp))
        temp[1] = 0
        i += 1
    return factorisation

def num_divisors(n, primes):
    if n < 2:
        return 1
    factorisation = prime_factorisation(n, primes)
    return reduce(lambda x,y:  x*y, [x[1]+1 for x in factorisation])


primes = []
M = 2000000
for i in range(2, M):
    print(i, "\r", end="")
    if is_prime_basic(i):
        primes.append(i)
print()

s , i = 0, 1
nd = num_divisors(s, primes)
while nd < 500:
    s += i
    nd = num_divisors(s, primes)
    i += 1
    print(i - 1, nd, "\r", end="")

print("\n", s)

File: cpp/12.cpp
#include <iostream>
#include <vector>

using namespace std;

bool is_prime(int n){
    if (n <= 1)
    {
        return false;
    } else if (n <= 3)
    {
        return true;
    } else if (n % 2 == 0 || n % 3 == 0)
    {
        return false;
    }

    int i =5;

    while (i * i <= n)
    {
        if (n % i == 0 || n % (i+2) == 0)
        {
            return false;
        }
        i += 6;
    }
    return true;
}

vector<vector<int>> prime_factorisation(int n, vector<int> primes){
    vector<vector<int>> factorisation{};
    int i{0};
    vector<int> temp{0, 0};
    while (n != 1)
    {
        int p = primes[i];
        while (n % p == 0)
        {
            temp[0] = p;
            temp[1] = temp[1] + 1;
            n /= p;
        }
        factorisation.push_back(temp);
        temp[1] = 0;
        i += 1;
    }
    return factorisation;
}

int num_divisors(int n, vector<int> primes){
    if (n < 2)
    {
       return 1;
    }
    vector<vector<int>> factorisation = prime_factorisation(n, primes);
    int num{1};
    for (int i = 0; i < factorisation.size(); i++)
    {
        num *= factorisation[i][1] + 1;
    }
    
    return num;
}

int main() {
    int M = 2000000;
    vector<int> primes{};

    for (int i = 2; i < M; i++)
    {
        if (is_prime(i))
        {
            primes.push_back(i);
        }
        
    }
    
    int s{0}, i{1};
    int nd = num_divisors(s, primes);

    while (nd < 500)
    {
        s += i;
        nd = num_divisors(s, primes);
        i++;
    }
    

    cout << s << endl;

    return 0;
}